provider/equinix/environ.go

```diff
@@ -0,0 +1,665 @@
+		if !deviceTags.Intersection(cpTags).IsEmpty() {
+			toReturn = append(toReturn, &equnixDevice{e, &cp})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func (e *environ) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+}
+
+func (e *environ) Config() *config.Config {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfg.config
+}
+
+func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
+}
+
+func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-is-controller": "true", "juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return nil, err
+	}
+	instanceIDs := make([]instance.Id, len(insts))
+	for _, i := range insts {
+		instanceIDs = append(instanceIDs, i.Id())
+	}
+	return instanceIDs, nil
+}
+
+func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
+	return nil
+}
+
+func (e *environ) Destroy(ctx context.ProviderCallContext) error {
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, inst := range insts {
+		if _, err = e.equnixClient.Devices.Delete(string(inst.Id()), true); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return common.Destroy(e, ctx)
+}
+
+func (e *environ) DestroyController(ctx context.ProviderCallContext, controllerUUID string) error {
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return err
+	}
+
+	for _, inst := range insts {
+		_, err = e.equnixClient.Devices.Delete(string(inst.Id()), true)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return e.Destroy(ctx)
+}
+
+func (e *environ) InstanceTypes(context.ProviderCallContext, constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
+	i := envinstance.InstanceTypesWithCostMetadata{}
+	instances, err := e.supportedInstanceTypes()
+	if err != nil {
+		return i, errors.Trace(err)
+	}
+
+	i.InstanceTypes = instances
+	return i, nil
+}
+
+func (e *environ) Instances(ctx context.ProviderCallContext, ids []instance.Id) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+
+	tags := set.NewStrings("juju-model-uuid=" + e.Config().UUID())
+
+	for _, id := range ids {
+		//TODO handle case when some of the instanes are missing
```

Comment:
This needs to be addressed in this PR. 

First, you need to pre-allocate a slice for the return values (instead of appending as the code below does). Then change the for loop (L195) to also use an index and use it to populate the appropriate index in the return value. Also, define a counter for missing instances outside the loop. If the client explicitly reports that a device is not found (i.e. **not** another kind of error occurred) just increment the counter and continue.

L206 can then be changed to something like this:

```
if missingInstanceCount > 0 {
   if missingInstanceCount == len(toReturn) {
      return nil, environs.ErrNoInstances
    } 

    return toReturn, environs.ErrPartialInstances
}
```